src/circular_queue.cpp

```diff
@@ -39,7 +39,7 @@ void CircularQueue::push(char *elm, int size) {
 	while(true) {
 		std::unique_lock<std::mutex> lk(write_lock);
 		// printf("CQ: push: wait on not-full. full() = %s\n", (full())? "true" : "false");
-		cirq_full.wait(lk, [this]{return !full();});
+		cirq_full.wait_for(lk, std::chrono::seconds(2), [this]{return !full();});
```

Comment:
hold on, why does pure wait cause an error?

Reply:
In some small number of cases "something" was happening which made both the push and peek wait on the queue because of the condition variables. Then the entire process ground to a halt. The wait_for enforces that the longest this could possibly happen for is 2 seconds.

Reply:
Any additional cost because of wait_for should be negligible due to the longer timeout.